### ISSUE TITLE
Some of the Got issues that could be applied to Ky

### ISSUE THREAD

**Issue by szmarczak** [sindresorhus/ky#129]
[Storing request specific context in Got](https://github.com/sindresorhus/got/issues/740)

[Pagination support](https://github.com/sindresorhus/got/issues/722)

[Ambiguous URLs](https://github.com/sindresorhus/got/issues/783)

Also, we could possibly add these hooks:

 - [ ] `beforeRedirect`
 - [ ] `beforeRetry`
 - [ ] `beforeError`

Also see https://github.com/sindresorhus/ky/pull/128#issuecomment-485771938 (`mutableDefaults`)

**Comment by sholladay** [comment]
Ky has a `json` option, whereas Got automatically detects objects in the `body` option an acts accordingly. Would be nice to make a decision on which one is better and have the same API. I think I lean towards the latter. I guess the idea with the `json` option was to simplify the implementation, but now we have to check for whether `body` was used as well and throw an error. So maybe it's a wash?

**Comment by szmarczak** [comment]
> Ky has a json option, whereas Got automatically detects objects in the body option an acts accordingly.

Got has [a json option](https://github.com/sindresorhus/got/blob/master/readme.md#json) too. `body` is meant for strings, buffers, streams and form-data instances only. This is beacuse we don't know if we should call `.toString()` or `JSON.stringify(...)`.

**Comment by sholladay** [comment]
[This example](https://github.com/sindresorhus/got/blob/master/readme.md#json-mode) may be out of date, then?

**Comment by szmarczak** [comment]
Indeed. I think @sindresorhus has told me about it some time ago and I forgot to change that :P

**Comment by RemyLivewall** [comment]
A before error hook would be a nice addition to handle timeouts or specific errors that can be handled in the same way to prevent unnecessary duplication of error handling code.

**Comment by cassiechew** [comment]
100% agree, adding pagination support would be quite nice.

**Comment by sholladay** [comment]
I don't think `beforeRedirect` is actionable here. While browser `fetch()` technically has a `redirect: 'manual'` mode, it causes an [opaque response](https://fetch.spec.whatwg.org/#concept-filtered-response-opaque-redirect) to be returned. It's basically useless. There _is_ [`response.redirected`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirected), so we could probably support an `afterRedirect` hook. But it seems simpler to just check for that in `afterResponse` anyway.

Request-specific context was released in v1.13.0 via PR https://github.com/sindresorhus/ky/pull/772.

That leaves only pagination from this list.

**Comment by sindresorhus** [comment]
> 100% agree, adding pagination support would be quite nice.

Not something I plan built-in in Ky, but see: https://github.com/sindresorhus/ky/commit/2481afdbc2a6eefd468f2bf6c08c39a4ea5b3535
